### ISSUE TITLE
Jupyter cell debugging does not support "step into" the  third party library code with "justmycode:false"

### ISSUE THREAD

**Issue by ckxckx** [microsoft/vscode-jupyter#8146]
Jupyter cell debugging does not support "step into" the  third party library code with "justmycode:false" in launch.json.

Please fix it !

**Comment by rchiodo** [comment]
This is explicitly hardcoded here:

https://github.com/microsoft/vscode-jupyter/blob/350a17c56b64317788b269f618b24ef112806eaa/src/client/debugger/jupyter/debuggingManager.ts#L299

Likely because stepping into 3rd party code isn't going to do what you want. It will end up stepping through the display hooks for ipython and not the 3rd party library.

@int19h is there a way to add libs to the 'justMyCode' list?

**Comment by int19h** [comment]
I don't think we have anything documented for this, but pydevd exposes some environment variables to [control this](https://github.com/microsoft/debugpy/blob/655e35642c357b18344d589c1ae0751f90860060/src/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_filtering.py#L112-L151). Presumably you'd want the library to debug to be listed in `IDE_PROJECT_ROOTS`, but I'm not sure how this will interact with the default exclusion filters - @fabioz should know.

I wonder if perhaps the Jupyter extension should be using the exclusion filters on a lower level to specifically exclude all those display hooks, while leaving "justMyCode" setting to the user?

**Comment by rchiodo** [comment]
> I wonder if perhaps the Jupyter extension should be using the exclusion filters on a lower level to specifically exclude all those display hooks, while leaving "justMyCode" setting to the user?

That sounds interesting too. How do we set exclusion filters?

**Comment by int19h** [comment]
It's all in the same code linked above, so currently, also via env vars (or using pydevd APIs). But we can always plumb that API through debugpy as needed.

**Comment by roblourens** [comment]
Thanks, where does the justMyCode flag actually come in, is it implemented using those flags?

**Comment by speediedan** [comment]
In case it's helpful to others until this is addressed, I've been able to workaround the issue effectively using IDE_PROJECT_ROOTS as suggested but would note one needs to add the appropriate ipykernel temp files directory (/tmp by default on Linux) in addition to the third party library code in IDE_PROJECT_ROOTS, e.g.:
```bash
IDE_PROJECT_ROOTS=/tmp:/opt/miniconda/envs/<myenv>/lib/<python_ver>/site-packages/<target third party lib>
```
otherwise, the 3rd-party breakpoints will work but the other breakpoints you set in your notebook will stop working. Hope this helps!

**Comment by jhcsjcao** [comment]
Ok.  After about 20 hours of trying, I got vscode to step into to dependent python libraries during debugging. I running  Jupyter works too.

You have to use 'base' python interpreters - NOT python environments in vscode, and set 'justmycode' to false in vscode's launch.json.

My dependent library modules are in a subdiretory of the directory that contains my Jupyter notebook and is the first directory listed when I run 'os.sys.path'

Hope this helps!

**Comment by kbfifi** [comment]
I suppose what jhcsjcao mentions leads to the solution. For me that would probably leave me with fixing the directory list produced by os.sys.path:
['\\\\rock64n2.lan\\kbfifi\\jupyter',
 'd:\\Users\\kbfifi\\anaconda3\\python39.zip',
 'd:\\Users\\kbfifi\\anaconda3\\DLLs',
 'd:\\Users\\kbfifi\\anaconda3\\lib',
 'd:\\Users\\kbfifi\\anaconda3',
 '',
 'd:\\Users\\kbfifi\\anaconda3\\lib\\site-packages',
 'd:\\Users\\kbfifi\\anaconda3\\lib\\site-packages\\locket-0.2.1-py3.9.egg',
 'd:\\Users\\kbfifi\\anaconda3\\lib\\site-packages\\win32',
 'd:\\Users\\kbfifi\\anaconda3\\lib\\site-packages\\win32\\lib',
 'd:\\Users\\kbfifi\\anaconda3\\lib\\site-packages\\Pythonwin',
 'd:\\Users\\kbfifi\\anaconda3\\lib\\site-packages\\IPython\\extensions',
 'C:\\Users\\kbfifi\\.ipython',
 'C:\\Users\\kbfifi\\AppData\\Roaming\\Python\\Python39\\site-packages']

It doesn't look really wrong to me

Although Jupyter setting "Debug just my code" set to false I get 
`
Frame skipped from debugging during step-in.
Note: may have been skipped because of "justMyCode" option (default == true). Try setting "justMyCode": false in the debug configuration (e.g., launch.json).
`
What am I missing?

**Comment by rchiodo** [comment]
As stated here:
https://github.com/microsoft/vscode-jupyter/issues/8146#issuecomment-958443000

Just my code is not supported for stepping into cells. You have to set the environment variables like @int19h mentioned.

**Comment by fabioz** [comment]
The real fix here would be having: https://github.com/microsoft/debugpy/issues/869 (that way we'd be able to make a distinction when library code is called because of ipython or not).

**Comment by kbfifi** [comment]
@rchiodo Thanks for the hint.
I tried with from a DOS prompt:
`
set IDE_PROJECT_ROOTS="c:\tmp\JupyterNotebookTmp:d:\Users\kbfifi\anaconda3\python39.zip:d:\Users\kbfifi\anaconda3\DLLs:d:\Users\kbfifi\anaconda3\lib:d:\Users\kbfifi\anaconda3:d:\Users\kbfifi\anaconda3\lib\site-packages:d:\Users\kbfifi\anaconda3\lib\site-packages\locket-0.2.1-py3.9.egg:d:\Users\kbfifi\anaconda3\lib\site-packages\win32:d:\Users\kbfifi\anaconda3\lib\site-packages\win32\lib:d:\Users\kbfifi\anaconda3\lib\site-packages\Pythonwin:d:\Users\kbfifi\anaconda3\lib\site-packages\IPython\extensions:C:\Users\kbfifi\.ipython:C:\Users\kbfifi\AppData\Roaming\Python\Python39\site-packages"

code
`
No luck. @speediedan what would by the Windows for /tmp as mentioned earlier in the thread: "appropriate ipykernel temp files directory (/tmp by default on Linux) "

**Comment by fabioz** [comment]
@kbfifi the separator on Windows must be `;` instead of `:`.

**Comment by kbfifi** [comment]
@faboiz thanks for noting that. Unfortunately it wasn't enough to get the workaround working. What might still help is what speediedan mentions. Do you know what the Windows path is for /tmp in Linux?

Using IDE_PROJECT_ROOTS does seem to have effect although not the wished one:
Before starting to debug a cell, the breakpoint on the left of the code line is a filled red bullet. During execution the bullet becomes a grey circle which seems invalid/disabled? Any ideas?

**Comment by fabioz** [comment]
@DonJayamanne @roblourens this issue doesn't seem to be fixed (the part related to `debugpy` is now done so that stepping should work in this scenario, but there's still work to be done so that users can actually configure the `justMyCode` setting in the vscode-jupyter side).

**Comment by smeunier-amd** [comment]
I would love a fix for this issue.

**Comment by saisravanreddy** [comment]
Resolution for this issue would be very helpful

**Comment by fabioz** [comment]
As a note, a workaround while a proper UI is actually done would be changing the code locally to enable it...

This can be done as a 1-line change in the method:

`pydevd.PyDB.set_use_libraries_filter`

To find the proper place you can run (under the debugger):

```
import pydevd
print(pydevd.__file__)
```

Then open the related file, search the method `set_use_libraries_filter` (i.e.: https://github.com/microsoft/debugpy/blob/v1.6.3/src/debugpy/_vendored/pydevd/pydevd.py#L1359)

And then locally change the variable `use_libraries_filter` to the value you want (if you want `justMyCode:false`, set `use_libraries_filter=False` as the first thing in that method so that the parameter value is overridden to the value you want).

**Comment by farfalk** [comment]
> As a note, a workaround while a proper UI is actually done would be changing the code locally to enable it...
> 
> This can be done as a 1-line change in the method:
> 
> `pydevd.PyDB.set_use_libraries_filter`
> 
> To find the proper place you can run (under the debugger):
> 
> ```
> import pydevd
> print(pydevd.__file__)
> ```
> 
> Then open the related file, search the method `set_use_libraries_filter` (i.e.: https://github.com/microsoft/debugpy/blob/v1.6.3/src/debugpy/_vendored/pydevd/pydevd.py#L1359)
> 
> And then locally change the variable `use_libraries_filter` to the value you want (if you want `justMyCode:false`, set `use_libraries_filter=False` as the first thing in that method so that the parameter value is overridden to the value you want).
> 
> Oh, and also an environment variable `PYDEVD_IPYTHON_COMPATIBLE_DEBUGGING=1` must be set (or you can manually edit the value in `pydevd_constants.py` which is right next to `pydevd.py` too if setting the environment variable prior to importing the debugger is hard in the use case).

I confirm this works with Windows 10, WSL2 Ubuntu 22.04, VSCode v1.71.1, python 3.10.4, python extension v2022.14.0. @fabioz you saved my day :')

**Comment by jpauwels** [comment]
> manually edit the value in `pydevd_constants.py` which is right next to `pydevd.py`

I got it working under macOS 12.6, Python 3.9.14 too, but I found the constants file at `pydevd/_pydevd_bundle/pydevd_constants.py` instead of the mentioned `pydevd/pydevd_constants.py` (in case it helps someone else). Many thanks!

**Comment by fabioz** [comment]
> I got it working under macOS 12.6, Python 3.9.14 too, but I found the constants file at `pydevd/_pydevd_bundle/pydevd_constants.py` instead of the mentioned `pydevd/pydevd_constants.py` (in case it helps someone else). Many thanks!

Thanks for the note. I've updated the original comment to point the proper location.

**Comment by roblourens** [comment]
Hey @fabioz, wondering whether you can give some guidance on what I need to adopt this? We are already setting the environment var, and if I send justMyCode:false, it works to step into library code, but I also keep stepping back out to interactiveshell.py after each line when stepping through the cell code.

<img width="763" alt="image" src="https://user-images.githubusercontent.com/323878/195490031-6de0742f-25bc-409e-bbb4-06a361662d18.png">

That seems to be a result of what you say about ipython creating a frame for each line, in https://github.com/microsoft/debugpy/issues/869. What do I need to do to avoid this?

**Comment by fabioz** [comment]
@roblourens I'm not sure, setting `PYDEVD_IPYTHON_COMPATIBLE_DEBUGGING=1` along with `justMyCode`: `false` should suffice. If it's still getting to `interactiveshell.py` after each line with those changes I probably need to investigate on the debugger side (as the setting related to `PYDEVD_IPYTHON_COMPATIBLE_DEBUGGING` is done specifically to overcome that issue).

Can you also provide the version of `ipython`/`ipykernel` you're using? (so that I check using the same versions you're testing there...).

**Comment by roblourens** [comment]
debugpy: 1.6.3
ipykernel: 6.15.2
ipython: 8.5.0


Am happy to help out with debugging as well if you tell me where I need to look.

**Comment by fabioz** [comment]
If you do a step-over, does it work (i.e.: does that happen only on a step-in)?

i.e.: ![vscode_record_notebook](https://user-images.githubusercontent.com/117621/195679111-0689ddfa-ad79-4064-a8a6-519de9012264.gif)

When I do a step in in the example above it does have some weird behavior that in stops more than once in the same place before going to the next line, but I haven't managed to get to the `interactiveshell.py`.

**Comment by roblourens** [comment]
Yeah, it happens on step over. Did you modify the extension to use justMyCode=false? Note that it currently is ignoring the setting for cell debugging (but respecting it for IW debugging, actually)

**Comment by fabioz** [comment]
> I double-checked that we are setting the environment variable, by executing some code it seems, but I also see this in the debugpy log:
> 
> `0.00s - Using PYDEVD_IPYTHON_COMPATIBLE_DEBUGGING: False`
> 
> does that mean that we set the env var too late, or will debugpy pick it up later?

Ok, that explains it... well, `pydevd` gets that flag when it's imported, but if you just import `debugpy` it won't import `pydevd` right away (it only does that afterwards when some internal API is used).

You can force `pydevd` to be imported with something as:

`import debugpy._vendored.force_pydevd`

**Comment by fabioz** [comment]
> Yeah, it happens on step over. Did you modify the extension to use justMyCode=false? Note that it currently is ignoring the setting for cell debugging (but respecting it for IW debugging, actually)

I did that following the steps in https://github.com/microsoft/vscode-jupyter/issues/8146#issuecomment-1235731548.

**Comment by fabioz** [comment]
As a note, it seems that `ipykernel` itself imports `debugpy`:

https://github.com/ipython/ipykernel/blob/main/ipykernel/debugger.py#L22

I think that `ipykernel` itself should set that environment variable when it's about to import `debugpy` / `pydevd`.

This also means that if you import `ipykernel` itself and then execute a command it's already too late (the shell that's used to start it should be configured prior to that).

**Comment by roblourens** [comment]
I'm actually kind of surprised that pydevd seems to be imported right away when the kernel starts, not when I start a debug session. @DonJayamanne do you know whether we cause debugpy/pydevd to be imported anywhere from our code anywhere before a debug session starts?

**Comment by fabioz** [comment]
> I'm actually kind of surprised that pydevd seems to be imported right away when the kernel starts, not when I start a debug session. @DonJayamanne do you know whether we cause debugpy/pydevd to be imported anywhere from our code anywhere before a debug session starts?

Seems like I was slow a few seconds for you to see my answer to that: https://github.com/microsoft/vscode-jupyter/issues/8146#issuecomment-1278052089 ;)

It's kind of a bummer that Python has eager imports vs lazy imports by default (meaning that IPyKernel loads its implementation of the debugger and debugpy by extension even if it's not going to be used in this case).

**Comment by roblourens** [comment]
Oops, yep thanks. But does ipykernel import that debugger.py right away?

**Comment by fabioz** [comment]
> Oops, yep thanks. But does ipykernel import that debugger.py right away?

Apparently yes: https://github.com/ipython/ipykernel/blob/main/ipykernel/ipkernel.py#L19

**Comment by fabioz** [comment]
Note: I also created an issue https://github.com/ipython/ipykernel/issues/1004 for `ipykernel` to do the right thing here.

**Comment by roblourens** [comment]
Thanks. I suppose that we can also actually spawn the kernel with the proper environment. But it really seems like the proper thing is for ipykernel to set this flag. Any reason we haven't already brought it up with them? Does Jupyter Lab know about this or have a solution to this?

**Comment by fabioz** [comment]
> Thanks. I suppose that we can also actually spawn the kernel with the proper environment. But it really seems like the proper thing is for ipykernel to set this flag. Any reason we haven't already brought it up with them? Does Jupyter Lab know about this or have a solution to this?

Well, I just brought it up in that ticket (some ipython devs were on https://github.com/microsoft/debugpy/issues/869, so I thought they were up to date on this, but maybe they didn't notice it).

**Comment by roblourens** [comment]
Don pushed a change in https://github.com/microsoft/vscode-jupyter/issues/11682 to spawn the kernel with the environment variable set, and I confirmed that with that change, justMyCode can work perfectly. So we have that for local kernels, and we just have to keep on top of https://github.com/ipython/ipykernel/issues/1004 as well. Thanks @fabioz and @DonJayamanne!

**Comment by roblourens** [comment]
Hm, this is actually not working for the Interactive Window. When I start debugging, I always start at that same line in interactiveshell.py as above https://github.com/microsoft/vscode-jupyter/issues/8146#issuecomment-1276968439 and can't hit a breakpoint in the cell code. I confirmed that the fix from yesterday is applying to the IW so I see `Using PYDEVD_IPYTHON_COMPATIBLE_DEBUGGING: True`. Can't tell why the IW would be different from the notebook here, any idea @DonJayamanne?

**Comment by roblourens** [comment]
I figured out that this happens with the `breakpoint()` statement, we inject this into the first line of the code run from the IW, and every time this runs, it breaks in interactiveshell.py. Happens if that statement is anywhere in the cell, when the above environment var is set.

@fabioz do you expect `breakpoint()` to work correctly here?

**Comment by roblourens** [comment]
We can probably get that working using a normal breakpoint. I'll open a new issue for that.

**Comment by fabioz** [comment]
> I figured out that this happens with the `breakpoint()` statement, we inject this into the first line of the code run from the IW, and every time this runs, it breaks in interactiveshell.py. Happens if that statement is anywhere in the cell, when the above environment var is set.
> 
> @fabioz do you expect `breakpoint()` to work correctly here?

When you use `breakpoint()` what you're doing is asking the debugger to stop as soon as possible in the first opportunity, which isn't really what you want in this case (because the first opportunity is in the library code).

What you can do instead in this case is ask the debugger to stop just in user code.

The api for this is not as simple as just `breakpoint()` though. 

If you know that you want to stop just at the main thread you can do something as:

```
import debugpy
debugpy.trace_this_thread()

from _pydevd_bundle.pydevd_api import PyDevdAPI
PyDevdAPI().stop_on_entry()
```

Note that this works just for the main thread (because `stop_on_entry()` makes the configuration for the main thread). We can create a better API for this use case though (maybe a keyword argument to `breakpoint()` like `breakpoint(stop_on_user_code=True)` would be good?)

**Comment by sachinruk** [comment]
To anyone else coming here looking for an answer 
1. You need the pre-release version of Jupyter extensions (as of 12th Nov 2022)
<img width="615" alt="image" src="https://user-images.githubusercontent.com/1410927/201473525-ff051734-9f01-4538-8d54-853d0a8ffaa2.png">

2. Make sure you have the following in your local `.vscode/settings.json` file.
```
{
    "jupyter.debugJustMyCode": false,
}
```


This takes in the solution of [the highest rated answer above](https://github.com/microsoft/vscode-jupyter/issues/8146#issuecomment-1235731548) and puts it into actual extension.

**Comment by nofreewill42** [comment]
holy jesus christ!
@sachinruk 
You're my savior man!!

**Comment by nofreewill42** [comment]
This is lifechanging!

**Comment by nofreewill42** [comment]
Alright it's still not the best it could be, but still:
I don't have to create a new python script that I can debug as I wish just because I've started out with a jupyter notebook where doing so is not really possible.

**Comment by nofreewill42** [comment]
Okay, this is actually falling apart.... oh jeez I thought I'm saved but this is getting uglier and uglier : (

**Comment by nofreewill42** [comment]
OH YES!!!
This is actually working maaan!!
(just don't soft link your ~/venv to ./.venv in the project directory lol....)

I'm saaaaaaaaaaaaved !!!! :D
I feel like the cows that are released to freedom for the first time in their lives:
https://www.youtube.com/watch?v=jQQTmuOEPLU&ab_channel=TomPembertonFarmLife&t=296

**Comment by userfirstname** [comment]
My experience is that it doesn't work if the module i am using is not in my current folder.
Graphically the two situation i have are:

Folder1: -notebook.ipynb  (where i import module.py)
              - module.py
          
Folder1: -notebook.ipynb  (where i import  module.py)
               Folder 2 : -module.py

The first one stops on breakpoints placed in module.py  when i use the Debug Cell. In the second one DebugCell run module.py without stoping on the breakpoints. 

I've tried with pre-released and released version and checked in the .vscode/settings.json that the right setting was used. I also tried to add the path to Folder2 in the setting file with "Python.analysis.extraPaths" with no luck. 

Thanks in advance for any help.
